File: client_app/lib/utils/route_gate.dart
import 'package:timezone/timezone.dart' as tz;

/// Matches server `computeGate` in `src/lib/route-gate.ts` (Asia/Manila).
enum RouteGate { onTime, early, late }

int _scheduledDayOrder(String recurrenceDay) {
  switch (recurrenceDay.toLowerCase()) {
    case 'sunday':
      return 0;
    case 'monday':
      return 1;
    case 'tuesday':
      return 2;
    case 'wednesday':
      return 3;
    case 'thursday':
      return 4;
    case 'friday':
      return 5;
    case 'saturday':
      return 6;
    default:
      return 0;
  }
}

/// Map Dart [DateTime.weekday] to server weekday order (Sun=0 .. Sat=6).
int _todayOrderFromWeekday(int dartWeekday) {
  return dartWeekday == DateTime.sunday ? 0 : dartWeekday;
}

RouteGate computeRouteGate({
  required String recurrenceDay,
  required int startHour,
  required int endHour,
}) {
  final int scheduled = _scheduledDayOrder(recurrenceDay);
  final tz.Location manila = tz.getLocation('Asia/Manila');
  final tz.TZDateTime now = tz.TZDateTime.now(manila);
  final int today = _todayOrderFromWeekday(now.weekday);
  final int hour = now.hour;

  if (today < scheduled) return RouteGate.early;
  if (today > scheduled) return RouteGate.late;
  if (hour < startHour) return RouteGate.early;
  if (hour >= endHour) return RouteGate.late;
  return RouteGate.onTime;
}

/// YYYY-MM-DD in Asia/Manila (for `routes.route_date`).
String routeDateInManila() {
  final tz.Location manila = tz.getLocation('Asia/Manila');
  final tz.TZDateTime now = tz.TZDateTime.now(manila);
  final String m = now.month.toString().padLeft(2, '0');
  final String d = now.day.toString().padLeft(2, '0');
  return '${now.year}-$m-$d';
}

File: src/lib/route-gate.ts
export type RouteGate = "on_time" | "early" | "late";

const DAY_ORDER: Record<string, number> = {
  sunday: 0,
  monday: 1,
  tuesday: 2,
  wednesday: 3,
  thursday: 4,
  friday: 5,
  saturday: 6,
};

export type TemplateGateInput = {
  recurrence_day: string;
  start_hour: number;
  end_hour: number;
};

/**
 * Compare scheduled window vs "now" in Asia/Manila (route ops TZ for TrashMap PH).
 */
export function computeGate(template: TemplateGateInput, now: Date): RouteGate {
  const scheduledDay = DAY_ORDER[template.recurrence_day.toLowerCase()];
  if (scheduledDay === undefined) {
    return "on_time";
  }

  const formatter = new Intl.DateTimeFormat("en-US", {
    timeZone: "Asia/Manila",
    weekday: "long",
    hour: "numeric",
    hour12: false,
  });
  const parts = formatter.formatToParts(now);
  const weekdayName = (parts.find((p) => p.type === "weekday")?.value ?? "sunday").toLowerCase();
  const hourRaw = parts.find((p) => p.type === "hour")?.value ?? "0";
  const today = DAY_ORDER[weekdayName] ?? 0;
  const hour = Number.parseInt(hourRaw, 10);
  const safeHour = Number.isFinite(hour) ? hour : 0;

  if (today < scheduledDay) return "early";
  if (today > scheduledDay) return "late";

  if (safeHour < template.start_hour) return "early";
  if (safeHour >= template.end_hour) return "late";
  return "on_time";
}

/** Calendar date YYYY-MM-DD in Asia/Manila (for route_date column). */
export function routeDateInManila(d: Date = new Date()): string {
  return d.toLocaleDateString("en-CA", { timeZone: "Asia/Manila" });
}
